File: v18_decorator/java/CompanyNotification.java
package v18_decorator.java;

// Notifier interface
interface Notifier {
    void send(String message);
}

// Concrete Component
class EmailNotifier implements Notifier {
    @Override
    public void send(String message) {
        System.out.println("Sending Email message: " + message);
    }
}

// Decorator abstract class
abstract class NotifierDecorator implements Notifier {
    protected Notifier notifier;

    public NotifierDecorator(Notifier notifier) {
        this.notifier = notifier;
    }

    @Override
    public void send(String message) {
        notifier.send(message);
    }
}

// Concrete Decorator classes
class WeChatDecorator extends NotifierDecorator {
    public WeChatDecorator(Notifier notifier) {
        super(notifier);
    }

    @Override
    public void send(String message) {
        super.send(message);
        System.out.println("Sending WeChat message: " + message);
    }
}

class MobileDecorator extends NotifierDecorator {
    public MobileDecorator(Notifier notifier) {
        super(notifier);
    }

    @Override
    public void send(String message) {
        super.send(message);
        System.out.println("Sending Mobile message: " + message);
    }
}

// Client
public class CompanyNotification {
    public static void main(String[] args) {
        String message = "Hello, this is a notification!";
        
        // Sending notifications using different combinations
        Notifier notifier = new EmailNotifier();
        notifier = new WeChatDecorator(notifier);
        notifier = new MobileDecorator(notifier);
        notifier.send(message);
    }
}

File: v18_decorator/python/company_notification.py
from abc import ABC, abstractmethod

# Notifier interface
class Notifier(ABC):
    @abstractmethod
    def send(self, message: str):
        pass

# Concrete Component
class EmailNotifier(Notifier):
    def send(self, message: str):
        print(f"Sending Email message: {message}")

# Decorator abstract class
class NotifierDecorator(Notifier):
    def __init__(self, notifier: Notifier):
        self.notifier = notifier

    def send(self, message: str):
        self.notifier.send(message)

# Concrete Decorator classes
class WeChatDecorator(NotifierDecorator):
    def __init__(self, notifier: Notifier):
        super().__init__(notifier)

    def send(self, message: str):
        super().send(message)
        print(f"Sending WeChat message: {message}")

class MobileDecorator(NotifierDecorator):
    def __init__(self, notifier: Notifier):
        super().__init__(notifier)

    def send(self, message: str):
        super().send(message)
        print(f"Sending Mobile message: {message}")

# Client
if __name__ == "__main__":
    message = "Hello, this is a notification!"

    # Sending notifications using different combinations
    notifier = EmailNotifier()
    notifier = WeChatDecorator(notifier)
    notifier = MobileDecorator(notifier)
    notifier.send(message)


